lib/Sema/LookupVisibleDecls.cpp

```diff
@@ -629,50 +631,66 @@ static void lookupVisibleMemberDeclsImpl(
     }
   }
 
-  llvm::SmallPtrSet<ClassDecl *, 8> Ancestors;
-  do {
-    NominalTypeDecl *CurNominal = BaseTy->getAnyNominal();
-    if (!CurNominal)
-      break;
-
-    synthesizeMemberDeclsForLookup(CurNominal, CurrDC);
+  const auto synthesizeAndLookupTypeMembers = [&](NominalTypeDecl *NTD) {
+    synthesizeMemberDeclsForLookup(NTD, CurrDC);
 
     // Look in for members of a nominal type.
     lookupTypeMembers(BaseTy, BaseTy, Consumer, CurrDC, LS, Reason);
+  };
+
+  llvm::SmallPtrSet<ClassDecl *, 8> Ancestors;
+  {
+    const auto NTD = BaseTy->getAnyNominal();
+    if (NTD == nullptr)
+      return;
+
+    synthesizeAndLookupTypeMembers(NTD);
     lookupDeclsFromProtocolsBeingConformedTo(BaseTy, Consumer, LS, CurrDC,
                                              Reason, Visited);
-    // If we have a class type, look into its superclass.
-    auto *CurClass = dyn_cast<ClassDecl>(CurNominal);
 
+    const auto CD = dyn_cast<ClassDecl>(NTD);
     // FIXME: We check `getSuperclass()` here because we'll be using the
     // superclass Type below, and in ill-formed code `hasSuperclass()` could
     // be true while `getSuperclass()` returns null, because the latter
     // looks for a declaration.
-    if (CurClass && CurClass->getSuperclass()) {
+    //
+    // If we have a superclass, switch state and look into the
+    // inheritance chain.
+    if (CD && CD->getSuperclass()) {
+      Ancestors.insert(CD);
+
+      Reason = getReasonForSuper(Reason);
+      BaseTy = CD->getSuperclass();
+
+      LS = LS.withOnSuperclass();
+      if (CD->inheritsSuperclassInitializers())
+        LS = LS.withInheritsSuperclassInitializers();
+    } else {
+      return;
+    }
+  }
+
+  // Look into the inheritance chain.
+  while (const auto CurClass = cast_or_null<ClassDecl>(
+                                   BaseTy->getAnyNominal())) {
```

Comment:
How about
```
do {
  const auto CurClass = BaseTy->getAnyNominal();
  assert(CurClass && "getSuperclass() returned non-nominal type");

  ...


  BaseTy = CurClass->getSuperclass();
  if (!CurClass->inheritsSuperclassInitializers())
    LS = LS.withoutInheritsSuperclassInitializers();
} while (BaseTy);
```

Reply:
Looks neat, but I think a `cast<>` is enough to indicate we have either tripped on something other than a class or a class type that is not bound (both are bad).

Reply:
Ah, sorry. we need to cast to `ClassDecl`. But anyway, here, `BaseType` is returned from `getSuperclass()` so `getAnyNominal` must be a valid `ClassDecl`.

Reply:
So `getClassOrBoundGenericClass()` should work I guess.